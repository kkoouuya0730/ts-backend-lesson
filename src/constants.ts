export const USER_NOT_FOUND = "User not found";
export const INVALID_USER_ID = "Invalid user ID";
export const INVALID_USER_INPUT = "Invalid user input";
export const USER_ALREADY_EXISTS = "User already exits";
