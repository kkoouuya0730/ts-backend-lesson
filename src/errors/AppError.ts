export class AppError extends Error {
  constructor(public statusCode: number, public message: string) {
    super(message);
    this.statusCode = statusCode;
    this.message = message;

    Object.setPrototypeOf(this, new.target.prototype);
  }
}
