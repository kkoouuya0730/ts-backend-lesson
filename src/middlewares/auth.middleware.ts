import { Request, Response, NextFunction } from "express";
import jwt from "jsonwebtoken";

type JwtPayload = {
  id: number;
  email: string;
};

export const authMiddleware = (req: Request, res: Response, _next: NextFunction) => {
  const authHeader = req.headers["authorization"];
  if (!authHeader) {
    return res.status(401).json({ error: "Authorization header missing" });
  }

  const token = authHeader.split(" ")[1];
  if (!token) {
    return res.status(401).json({ error: "Token missing" });
  }

  try {
    const secret = process.env.JWT_SECRET!;
    const decoded = jwt.verify(token, secret) as JwtPayload;
    req.user = decoded;
  } catch (error) {
    return res.status(401).json({ error: "Invalid or expired token" });
  }
};
