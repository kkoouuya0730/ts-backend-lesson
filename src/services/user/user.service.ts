import prisma from "../../models/prisma";
import { UserInput } from "../../validation";

export const getUsers = async () => {
  const users = await prisma.user.findMany();
  return users.map(({ passwordHash, ...publicUser }) => publicUser);
};

export const getUser = async (id: number) => {
  const user = await prisma.user.findUnique({
    where: {
      id,
    },
  });
  if (!user) return null;
  const { passwordHash, ...publicUser } = user;
  return publicUser;
};

export const createUser = async (data: UserInput) => {
  return await prisma.user.create({
    data,
  });
};

export const updateUser = async (id: number, name: string) => {
  const updatedUser = await prisma.user.update({
    where: {
      id,
    },
    data: {
      name,
    },
  });
  const { passwordHash, ...publicUser } = updatedUser;
  return publicUser;
};

export const deleteUser = async (id: number) => {
  await prisma.user.delete({
    where: {
      id,
    },
  });
};
