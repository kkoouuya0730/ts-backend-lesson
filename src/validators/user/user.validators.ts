import { ZodSafeParseResult } from "zod";
import { userIdSchema, UserInput, userInputSchema } from "../../validation";

/**
 * 例外を投げる関数を限定するという意図のもと、
 * parse系関数では例外を投げず、呼び出しもとで例外処理を行う
 */

export const parseUserId = (
  id: string | undefined
): ZodSafeParseResult<{
  id: number;
}> => {
  const parsedUserIdResult = userIdSchema.safeParse({ id });
  return parsedUserIdResult;
};

export const parseUserInput = (
  input: UserInput | undefined
): ZodSafeParseResult<{
  name: string;
}> => {
  const parsedInput = userInputSchema.safeParse(input);
  return parsedInput;
};

export const parseCreateUserInput = (
  input: UserInput | undefined
): ZodSafeParseResult<{
  name: string;
}> => {
  const parsedInput = userInputSchema.safeParse(input);
  return parsedInput;
};
