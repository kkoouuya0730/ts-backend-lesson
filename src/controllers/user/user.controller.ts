import * as userService from "../../services/user/user.service";
import { Request, Response, NextFunction } from "express";
import { AppError } from "../../errors/AppError";
import { INVALID_USER_ID, INVALID_USER_INPUT, USER_NOT_FOUND } from "../../constants";
import { userIdSchema, userInputSchema } from "../../validation";

export const getUsersHandler = async (_req: Request, res: Response, next: NextFunction) => {
  try {
    const users = await userService.getUsers();
    res.json(users);
  } catch (error) {
    next(error);
  }
};

export const getUserHandler = async (req: Request, res: Response, next: NextFunction) => {
  const parsedUserIdResult = userIdSchema.safeParse({ id: req.params.id });
  console.log(parsedUserIdResult);
  if (!parsedUserIdResult.success) {
    throw new AppError(400, INVALID_USER_ID, parsedUserIdResult.error.issues);
  }

  const userId = parsedUserIdResult.data.id;

  try {
    const user = await userService.getUser(userId);
    if (!user) {
      throw new AppError(404, USER_NOT_FOUND);
    }
    res.json(user);
  } catch (error) {
    next(error);
  }
};

export const updateUserHandler = async (req: Request, res: Response, next: NextFunction) => {
  const parsedUserIdResult = userIdSchema.safeParse({ id: req.params.id });

  if (!parsedUserIdResult.success) {
    throw new AppError(400, INVALID_USER_ID, parsedUserIdResult.error.issues);
  }

  const userId = parsedUserIdResult.data.id;

  const parsedInput = userInputSchema.safeParse(req.body);
  if (!parsedInput.success) {
    throw new AppError(400, INVALID_USER_INPUT, parsedInput.error.issues);
  }

  try {
    const user = await userService.updateUser(userId, parsedInput.data.name);
    res.json(user);
  } catch (error) {
    next(error);
  }
};

export const deleteUserHandler = async (req: Request, res: Response, next: NextFunction) => {
  const parsedUserIdResult = userIdSchema.safeParse({ id: req.params.id });
  if (!parsedUserIdResult.success) {
    throw new AppError(400, INVALID_USER_ID, parsedUserIdResult.error.issues);
  }

  const userId = parsedUserIdResult.data.id;

  try {
    await userService.deleteUser(userId);
    res.json({ message: "User deleted successfully" });
  } catch (error) {
    next(error);
  }
};
