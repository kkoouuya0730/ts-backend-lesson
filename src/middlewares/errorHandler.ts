import { Request, Response, NextFunction } from "express";
import { AppError } from "../errors/AppError";
import { ZodError } from "zod";
import axios from "axios";

export const errorHandler = (err: any, req: Request, res: Response, _next: NextFunction) => {
  console.error(err);

  axios
    .post(process.env.SLACK_WEBHOOK_URL!, {
      text: `❌ Error: ${err.message}\nPath: ${req.path}\nStack: ${err.stack}`,
    })
    .catch(() => {});

  if (err instanceof AppError) {
    return res.status(err.statusCode).json({
      error: {
        code: err.statusCode,
        message: err.message,
        details: err.details ?? null,
      },
    });
  }

  if (err instanceof ZodError) {
    return res.status(400).json({
      error: {
        code: 400,
        message: "Validation Error",
        details: err.issues.map((issue) => ({
          path: issue.path,
          message: issue.message,
        })),
      },
    });
  }

  const message = err.message || "Internal Server Error";

  res.status(500).json({ error: { code: 500, message } });
};
