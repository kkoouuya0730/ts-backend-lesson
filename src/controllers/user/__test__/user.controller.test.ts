import request from "supertest";
import express from "express";
import * as userController from "../../user/user.controller";
import * as userService from "../../../services/user/user.service";
import { AppError } from "../../../errors/AppError";
import { INVALID_USER_ID, INVALID_USER_INPUT, USER_NOT_FOUND } from "../../../constants";
import { errorHandler } from "../../../middlewares/errorHandler";
import { mockUser, mockUsers } from "../../../mocks/user.mock";

jest.mock("../../../services/user/user.service");

const app = express();
app.use(express.json());
app.get("/users", userController.getUsersHandler);
app.get("/users/:id", userController.getUserHandler);
// app.post("/users", userController.createUserHandler);
app.put("/users/:id", userController.updateUserHandler);
app.delete("/users/:id", userController.deleteUserHandler);
app.use(errorHandler);

describe("UserController", () => {
  afterEach(() => {
    jest.clearAllMocks();
  });
  describe("正常系", () => {
    test("getUsersHandler", async () => {
      (userService.getUsers as jest.Mock).mockResolvedValue(mockUsers);

      const res = await request(app).get("/users");

      expect(res.status).toBe(200);
      expect(res.body).toEqual(mockUsers);
      expect(userService.getUsers).toHaveBeenCalledTimes(1);
      expect(res.body).toMatchSnapshot();
    });

    test("getUserHandler", async () => {
      (userService.getUser as jest.Mock).mockResolvedValue(mockUser);

      const res = await request(app).get("/users/1");

      expect(res.status).toBe(200);
      expect(res.body).toEqual(mockUser);
      expect(userService.getUser).toHaveBeenCalledTimes(1);
      expect(res.body).toMatchSnapshot();
    });

    // TODO authに移動したのでそっちでテスト書く
    // test("createUserHandler", async () => {
    //   (userService.createUser as jest.Mock).mockResolvedValue(mockUser);

    //   const res = await request(app).post("/users").send({ name: "Alice" });

    //   expect(res.status).toBe(201);
    //   expect(res.body).toEqual(mockUser);
    //   expect(userService.createUser).toHaveBeenCalledTimes(1);
    //   expect(res.body).toMatchSnapshot();
    // });

    test("updateUserHandler", async () => {
      (userService.updateUser as jest.Mock).mockResolvedValue({
        id: 1,
        name: "Alice updated",
        email: "test@test.com",
      });

      const res = await request(app)
        .put("/users/1")
        .send({ name: "Alice updated", email: "test@test.com", passwordHash: "hogehoge" });

      expect(res.status).toBe(200);
      expect(res.body).toEqual({ id: 1, name: "Alice updated", email: "test@test.com" });
      expect(userService.updateUser).toHaveBeenCalledTimes(1);
      expect(res.body).toMatchSnapshot();
    });

    test("deleteUserHandler", async () => {
      (userService.deleteUser as jest.Mock).mockResolvedValue({ message: "User deleted successfully" });

      const res = await request(app).delete("/users/1");

      expect(res.status).toBe(200);
      expect(userService.deleteUser).toHaveBeenCalledTimes(1);
      expect(res.body).toMatchSnapshot();
    });
  });

  describe("異常系", () => {
    test("getUserHandler", async () => {
      (userService.getUser as jest.Mock).mockRejectedValue(new AppError(404, USER_NOT_FOUND));

      const res = await request(app).get("/users/1");

      expect(res.status).toBe(404);
      expect(res.body).toEqual({
        error: {
          code: 404,
          message: "User not found",
          details: null,
        },
      });
      expect(userService.getUser).toHaveBeenCalledTimes(1);
    });

    // test("createUserHandler", async () => {
    //   (userService.createUser as jest.Mock).mockRejectedValue(new AppError(400, INVALID_USER_INPUT));

    //   const res = await request(app).post("/users").send({ name: "" });

    //   expect(res.status).toBe(400);
    //   expect(res.body).toEqual({
    //     error: {
    //       code: 400,
    //       message: "Invalid user input",
    //       details: expect.any(Array),
    //     },
    //   });
    //   expect(userService.createUser).toHaveBeenCalledTimes(0);
    // });

    test("updateUserHandler request bodyが不正", async () => {
      (userService.updateUser as jest.Mock).mockRejectedValue(new AppError(400, INVALID_USER_INPUT));

      const res = await request(app).put("/users/1").send({ name: "" });

      expect(res.status).toBe(400);
      expect(res.body).toEqual({
        error: {
          code: 400,
          message: "Invalid user input",
          details: expect.any(Array),
        },
      });
      expect(userService.updateUser).toHaveBeenCalledTimes(0);
    });

    test("updateUserHandler path paramが不正", async () => {
      (userService.updateUser as jest.Mock).mockRejectedValue(new AppError(400, INVALID_USER_ID));

      const res = await request(app).put("/users/abc").send({ name: "Alice updated" });

      expect(res.status).toBe(400);
      expect(res.body).toEqual({
        error: {
          code: 400,
          message: "Invalid user ID",
          details: expect.any(Array),
        },
      });
      expect(userService.updateUser).toHaveBeenCalledTimes(0);
    });

    test("deleteUserHandler", async () => {
      (userService.deleteUser as jest.Mock).mockRejectedValue(new AppError(400, INVALID_USER_INPUT));

      const res = await request(app).delete("/users/abc");

      expect(res.status).toBe(400);
      expect(res.body).toEqual({
        error: {
          code: 400,
          message: "Invalid user ID",
          details: expect.any(Array),
        },
      });
      expect(userService.deleteUser).toHaveBeenCalledTimes(0);
    });
  });
});
