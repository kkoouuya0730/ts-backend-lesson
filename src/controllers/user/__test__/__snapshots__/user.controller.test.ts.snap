// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`UserController 正常系 deleteUserHandler 1`] = `
{
  "message": "User deleted successfully",
}
`;

exports[`UserController 正常系 getUserHandler 1`] = `
{
  "email": "test1@test.com",
  "id": 1,
  "name": "Alice",
  "passwordHash": "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
}
`;

exports[`UserController 正常系 getUsersHandler 1`] = `
[
  {
    "email": "test1@test.com",
    "id": 1,
    "name": "Alice",
    "passwordHash": "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
  },
  {
    "email": "test2@test.com",
    "id": 2,
    "name": "Bob",
    "passwordHash": "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
  },
]
`;

exports[`UserController 正常系 updateUserHandler 1`] = `
{
  "email": "test@test.com",
  "id": 1,
  "name": "Alice updated",
}
`;
